logdna/request_types.go

```diff
@@ -0,0 +1,200 @@
+			PAGERDUTY,
+			&diags,
+		)...,
+	)
+
+	view.Channels = append(
+		view.Channels,
+		*mapChannelsFromSchema(
+			d.Get("webhook_channel").([]interface{}),
+			WEBHOOK,
+			&diags,
+		)...,
+	)
+	return diags
+}
+
+func mapChannelsFromSchema(listEntries []interface{}, integration string, diags *diag.Diagnostics) *[]channelRequest {
+	var prepared interface{}
+	channelRequests := []channelRequest{}
+
+	if listEntries == nil {
+		return nil
+	}
+	for _, entry := range listEntries {
+		e := entry.(map[string]interface{})
+		prepared = nil
+		switch integration {
+		case EMAIL:
+			prepared = emailChannelRequest(e)
+		case PAGERDUTY:
+			prepared = pagerDutyChannelRequest(e)
+		case WEBHOOK:
+			prepared = webHookChannelRequest(e, diags)
+		default:
+			*diags = append(*diags, diag.Diagnostic{
+				Severity: diag.Error,
+				Summary:  "Cannot format integration channel for outbound request",
+				Detail:   fmt.Sprintf("Unrecognized integration: %s", integration),
+			})
+		}
+		if prepared == nil {
+			continue
+		}
+		channelRequests = append(channelRequests, prepared.(channelRequest))
+	}
+	return &channelRequests
+}
+
+func emailChannelRequest(s map[string]interface{}) channelRequest {
+	var emails []string
+	for _, email := range s["emails"].([]interface{}) {
+		emails = append(emails, email.(string))
+	}
+
+	c := channelRequest{
+		Emails:          emails,
+		Immediate:       s["immediate"].(string),
+		Integration:     EMAIL,
+		Operator:        s["operator"].(string),
+		Terminal:        s["terminal"].(string),
+		TriggerInterval: s["triggerinterval"].(string),
+		TriggerLimit:    s["triggerlimit"].(int),
+		Timezone:        s["timezone"].(string),
+	}
+
+	return c
+}
+
+func pagerDutyChannelRequest(s map[string]interface{}) channelRequest {
+	c := channelRequest{
+		Immediate:       s["immediate"].(string),
+		Integration:     PAGERDUTY,
+		Key:             s["key"].(string),
+		Operator:        s["operator"].(string),
+		Terminal:        s["terminal"].(string),
+		TriggerInterval: s["triggerinterval"].(string),
+		TriggerLimit:    s["triggerlimit"].(int),
+	}
+
+	return c
+}
+
+func webHookChannelRequest(s map[string]interface{}, diags *diag.Diagnostics) channelRequest {
+	headersMap := make(map[string]string)
+
+	for k, v := range s["headers"].(map[string]interface{}) {
+		headersMap[k] = v.(string)
+	}
+
+	c := channelRequest{
+		Headers:         headersMap,
+		Immediate:       s["immediate"].(string),
+		Integration:     WEBHOOK,
+		Operator:        s["operator"].(string),
+		Method:          s["method"].(string),
+		TriggerInterval: s["triggerinterval"].(string),
+		TriggerLimit:    s["triggerlimit"].(int),
+		URL:             s["url"].(string),
+		Terminal:        s["terminal"].(string),
+	}
+
+  if bodyTemplate := s["bodytemplate"].(string) ;bodyTemplate != "" {
+    var bt map[string]interface{}
+    // See if the JSON is valid, but don't use the value or it will double encode
+    err := json.Unmarshal([]byte(bodyTemplate), &bt)
```

Comment:
Here.  We could grab the value and assign it back to the data somehow.  Maybe.

Reply:
I messed with this, but it immediately got messy.  We'd have to pass around `*d` in the request models in order to mutate it.  What's worse though, is that the state needs a `string`, which would require an `Marshal(Unmarshal)` call.  That's not pleasing.

Using the diff comparator, we can at least compare the unmarshalled JSON, which is only 1 call.  Not only that, it will only run this when there's bodytemplate values, so it's efficient that way.